src/main/java/pink/zak/minestom/towerdefence/ui/spawner/TroopSpawnerUI.java

```diff
@@ -0,0 +1,180 @@
+            if (!p.equals(gameUser.getPlayer())) return;
+
+            // check if slots are within the tab range
+            if (this.tab != null && slot >= 9 && slot < 27) {
+                this.tab.onClick(slot - 9);
+                return;
+            }
+
+            // run the click handler
+            this.onClick(slot);
+        });
+    }
+
+    private void onClick(int slot) {
+        if (slot == 31 && this.mode == Mode.SEND) {
+            this.enterUpgradeMode();
+            return;
+        }
+
+        Optional<EnemyMob> optionalEnemyMob = this.mobStorage.getEnemyMobs().stream()
+                .filter(mob -> mob.getSlot() == slot)
+                .findFirst();
+        if (optionalEnemyMob.isEmpty()) return; // clicked on an empty slot
+
+        EnemyMob enemyMob = optionalEnemyMob.get();
+        int levelInt = this.gameUser.getMobLevel(enemyMob);
+        EnemyMobLevel level = enemyMob.getLevel(levelInt);
+
+        if (this.mode == Mode.SEND && level != null) attemptToSendMob(enemyMob);
+        if (this.mode == Mode.UPGRADE) setTab(new TroopUpgradeTab(this, this.gameUser, enemyMob));
+    }
+
+    private void attemptToSendMob(@NotNull EnemyMob mob) {
+        GameUser user = this.gameUser;
+
+        EnemyMobLevel level = mob.getLevel(user.getMobLevel(mob));
+        if (level == null) return; // the player attempted to send a mob that they don't own
+
+        // check if the player can afford to send the mob
+        if (user.getCoins() < level.getSendCost()) return;
+        if (!user.getQueue().canQueue(mob)) return;
+
+        // charge player for sending the mob
+        user.updateCoins(current -> current - level.getSendCost());
+
+        // add the mob to the queue
+        user.getQueue().queue(new QueuedEnemyMob(mob, level));
+        this.updateQueue(user.getQueue().getQueuedMobs());
+    }
+
+    private void enterUpgradeMode() {
+        this.mode = Mode.UPGRADE;
+        this.setTitle(UPGRADE_TITLE);
+        this.setItemStack(31, ItemStack.AIR);
+        this.setItemStack(35, ItemStack.AIR);
+    }
+
+    private void setTab(@NotNull TroopUpgradeTab tab) {
+        this.tab = tab;
+        updateSubInventory();
+    }
+
+    void updateSubInventory() {
+        if (this.tab == null) return;
+        for (int i = 9; i < 27; i++) setItemStack(i, this.tab.getItemStack(i - 9));
+    }
+
+    private @NotNull ItemStack createQueueItem(@NotNull Queue<QueuedEnemyMob> queue) {
+        List<ItemStack> items = new ArrayList<>();
+
+        EnemyMob currentlyTrackedMob = null;
+        int count = 0; // count of the current iterated mob
+        for (QueuedEnemyMob queuedMob : queue) {
+            EnemyMob mob = queuedMob.mob();
+            if (currentlyTrackedMob != mob) {
+                if (currentlyTrackedMob != null) {
+                    items.add(currentlyTrackedMob.getLevel(this.gameUser.getMobLevels().get(currentlyTrackedMob)).createPreviewItem().withAmount(count));
```

Comment:
Creating an item each iteration probably isn't a great idea, should probably count the number of mobs beforehand.